### ISSUE TITLE
Add CI configuration using NastyMPI

### ISSUE THREAD

**Issue by fuchsto** [dash-project/dash#52]
Some defects are only visible by chance, like missing barriers after local initialization.
I think we could drastically reduce false positives in CI builds by integrating NastyMPI.

**Comment by fmoessbauer** [comment]
Support for NastyMPI is provided in branch `feat-52-nastympi`. Currently NastyMPI has to be compiled by the user.

For the CI tests it would be nice if cmake pulls and builds it automatically if  specified (like google test). @rkowalewski can you provide a cmake build script for nastympi? That would make the process much easier.

**Comment by rkowalewski** [comment]
@fmoessbauer: Yes, I will do that.

**Comment by fmoessbauer** [comment]
@rkowalewski : Please check if I integrated NastyMPI correctly in DASH. How can I verify that the proxy is actually used?

**Comment by rkowalewski** [comment]
The original solution had two minor issues:
a) While we link statically we always use shared libraries. For this reason we should use the shared library version of Nasty-MPI. --> Fixed
b) The order in which libraries are listed during the link step mattes. Before the small fix, Nasty-MPI was simply added to the end of all the libraries. However, we have to link Nasty-MPI before the native MPI library. --> Fixed

On my local machine Nasty-MPI prints now debug output. I have checked it with a simple dash::array test.
@fmoessbauer All changes are now in the corresponding branch.

**Comment by fuchsto** [comment]
@rkowalewski Thanks!
But currenlty the NastyMPI-tests fail with:

~~~
/opt/dash/build-ci/20161128-223703--uuid-9d6d693f/Nasty/bin/dash/test/mpi/dash-test-mpi:
error while loading shared libraries: libnasty_mpi.so.0: cannot open shared object file:
No such file or directory
~~~

See logs in https://circleci.com/gh/dash-project/dash/551

**Comment by rkowalewski** [comment]
Is there any chance to see the linker commands on circle CI? I have checked my local build scripts and it works. Probably LD_LIBRARY_PATH, or similar, is different in the docker machine.

**Comment by fmoessbauer** [comment]
Hi, I don't think this is a circle CI specific problem, as the tests are run inside our docker containers. Just run the container locally as described in this [wonderful documentation](http://dash.readthedocs.io/en/latest/CI/#online-ci-providers)

**Comment by fmoessbauer** [comment]
The problem is quite simple. NastyMPI is not copied to DASH's install dir. Hence it is not found when running the CI script.

**Comment by fmoessbauer** [comment]
There is a bug in NastyMPI whereby it does not compile in gcc:
```
src/nasty_runtime/init.c:109:50: Fehler: ISO-C99 erfordert mindestens ein Argument für »...« in einem variadischen Makro [-Werror]
   log_info("Initialization successfully finished");
```

I temporary fixed that by removing the `-Werror` flag in branch `bug-gcc-macro`. @rkowalewski please have a look at this macro and fix it finally. However I am not sure if it is really desired to use `-Werror` in production builds.

For further details on this bug, see https://gcc.gnu.org/onlinedocs/cpp/Variadic-Macros.html

**Comment by rkowalewski** [comment]
@fmoessbauer Thanks! I am going to fix this. However, I agree on that we should not use the `-Werror` flag in productive builds. I will accordingly modify the Makefile.

**Comment by fuchsto** [comment]
@fmoessbauer @rkowalewski 

Before we can close this issue, some last questions:

**MPI shared windows**

If I remember correctly, MPI shared windows should be disabled for Nasty-MPI runs because, obviously, Nasty-MPI can't nastify `memcpy`. Right?

- Curiously, the local-to-global `copy_async`-test still fails with MPI shared windows enabled.
  Maybe we need `MPI_Win_fence` to synchronize private/public memory?

- If DASH is configured with MPI shared windows disabled, tests of DART one-sided communication fail.
  Looks like there are defects in some fundamental mechanics in `dart_mpi/dart_communication`, but I
  didn't notice anything suspicious in the code, yet.
  For example, this test call fails for `-DENABLE_SHARED_WINDOWS=OFF` and passes for `=ON`:
    
    ~~~bash
    LD_PRELOAD=~/opt/nasty-mpi/lib/libnasty_mpi.so \
      mpirun -n 2 ./bin/dash-test-mpi --gtest_filter="DARTOnesidedTest.GetBlockingSingleBlock"
    ~~~

**MPI memory model**

- Does NastyMPI assume the unified MPI memory model? Is there an option to toggle unified memory
  model nastiness in NastyMPI?

**Build config**

- Do we still have to configure DASH for Nasty-MPI via CMake option `ENABLE_NASTYMPI` or does 
  setting `LD_PRELOAD` suffice?

**Comment by fuchsto** [comment]
@fmoessbauer @rkowalewski 
MPI shared windows have been enabled in the Nasty-MPI build config which prevents Nasty-MPI from intercepting MPI function calls.

This is fixed in CI and build scripts on branch `feat-52-nastympi` now.

**Comment by fuchsto** [comment]
@rkowalewski 
Also, test even fail when executed on a single unit, see [CircleCI build 592](https://circleci.com/gh/dash-project/dash/592).
Found this rather surprising (note that it's `mpirun -n 1`):

~~~
[[ RUN    ]] [ 20161203-044535 ] timeout -s 15 -k 1m --foreground 1m  mpirun -map-by core  -n 1   /opt/dash/build-ci/20161203-044157--uuid-8678dd01/Nasty/bin/dash/test/mpi/dash-test-mpi --gtest_filter='DARTOnesidedTest.*'
[Nasty-MPI INFO] (src/nasty_runtime/init.c:109) Initialization successfully finished
[----------] 41 tests will be run.
[----------] run 3 out of 3 tests from DARTOnesidedTest
[  RUN     ] DARTOnesidedTest.GetBlockingSingleBlock
[  ERROR   ] [UNIT 0] in /opt/dash/dash/test/DARTOnesidedTest.cc:36
      Expected: expected
      Which is: 1001
To be equal to: local_array[l]
      Which is: -385875965
Unit 0
~~~

**Comment by fmoessbauer** [comment]
@fuchsto 
Addressing the "Build config" question above: Theoretically `LD_PRELOAD` is sufficient if MPI libs are linked dynamically. However even in this case the user has to provide a build of NastyMPI. If setting the flag above, cmake pulls NastyMPI from the official repo and builds it. Furthermore, DART is linked against it in the correct order, so that even static linked MPI libs are supported.

Normally `LD_PRELOAD` does not have to be set, as NastyMPI is already linked. However on CI (for whatever reason) `nastympi.so ` is not found, because the absolute path to the shared object is not set in the binary. You can check that:

```
ldd dash-test-mpi
[...]
libnasty_mpi.so.0 => /home/moessbauer/repos/github-dash/dash/build.nasty/nastympi/src/NastyMPIExternal/lib/libnasty_mpi.so.0 (0x00007ff6cf81b000)
[...]
```
On CI this path is empty / "not found".

**Comment by fuchsto** [comment]
@fmoessbauer @rkowalewski 

Finalization fails for OpenMPI + NastyMPI:

<https://638-45930061-gh.circle-artifacts.com/1/tmp/circle-junit.BjwvGVQ/openmpi/Nasty/logs/test_mpi.log>

Example:

~~~
[      OK  ] TeamTest.SplitTeamSync
[Nasty-MPI INFO] (src/nasty_runtime/init.c:121) Finalization successfully finished
[==========] 2 tests from 1 test cases ran. (5200 ms total)
[Nasty-MPI INFO] (src/nasty_runtime/init.c:121) Finalization successfully finished
[  PASSED  ] 2 tests passed
[Nasty-MPI INFO] (src/nasty_runtime/init.c:121) Finalization successfully finished
*** The MPI_Win_get_attr() function was called after MPI_FINALIZE was invoked.
*** This is disallowed by the MPI standard.
*** Your MPI job will now abort.
~~~

There is no error for MPICH and OpenMPI 2, though.

Update: Apparently the test suite has been killed as it timed out:

~~~
[[ RUN    ]] [ 20161207-074051 ] timeout -s 15 -k 3m --foreground 3m  mpirun -map-by core --allow-run-as-root -n 2   /opt/dash/build-ci/20161207-073619--uuid-f290664d/Nasty/bin/dash/test/mpi/dash-test-mpi --gtest_filter='TeamTest.*'
// --- SNIP ---
[[     OK ]] [ 20161207-074351 ] Tests passed, returned 0
~~~

**Comment by devreal** [comment]
Just for the records: Using Nasty MPI in CI seems to fail sporadically, see https://circleci.com/gh/dash-project/dash/731. Re-running the test suite succeeded.

**Comment by rkowalewski** [comment]
This is probably a progress problem in Open-MPI which appears even in the nightly build. I have still to figure it out.

**Comment by rkowalewski** [comment]
The build is now fixed. The errors were reproducible only by using Open-MPI 1.x or MPICH 3.x. It turned out that we did not properly finalize DASH before finalizing MPI. We have recently decoupled MPI Initialization from DASH Initialization. This is fixed now as well.

**Comment by fuchsto** [comment]
Fixed in #170
